presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,747 @@
+    private final AtomicLong totalRowsAdded = new AtomicLong();
+
+    private final OutputBuffers.OutputBufferId outputBufferId = new OutputBuffers.OutputBufferId(0);
+    private final Closer closer = Closer.create();
+
+    private static ExecutorService coreExecutor = newCachedThreadPool(daemonThreadsNamed("spooling-outputbuffer-%s"));
+    private static final Logger log = Logger.get(SpoolingOutputBuffer.class);
+
+    private final AtomicBoolean noMorePages = new AtomicBoolean();
+    private final AtomicLong memorySequenceId = new AtomicLong();
+    private final AtomicLong sequenceId = new AtomicLong();
+    private final AtomicLong startPage = new AtomicLong();
+    private final AtomicLong totalPagesRemaining = new AtomicLong();
+
+    @GuardedBy("this")
+    private final LinkedList<HandleInfo> handleInfoQueue = new LinkedList<>();
+
+    @GuardedBy("this")
+    private final Queue<SerializedPage> pages = new ArrayDeque<>();
+
+    @GuardedBy("this")
+    private PendingRead pendingRead;
+
+    public SpoolingOutputBuffer(
+            TaskId taskId,
+            String taskInstanceId,
+            OutputBuffers outputBuffers,
+            StateMachine<BufferState> state,
+            TempStorage tempStorage,
+            DataSize threshold)
+    {
+        requireNonNull(taskId, "taskId is null");
+        this.taskInstanceId = requireNonNull(taskInstanceId, "taskInstanceIs is null");
+        this.outputBuffers = requireNonNull(outputBuffers, "outputBuffers is null");
+        this.state = requireNonNull(state, "state is null");
+        this.tempStorage = requireNonNull(tempStorage, "tempStorage is null");
+        this.threshold = requireNonNull(threshold, "threshold is null");
+
+        context = new TempDataOperationContext(Optional.empty(), taskId.getQueryId().toString(), Optional.empty(), new Identity("spooling-buffer", Optional.empty()));
+
+        closer.register(coreExecutor::shutdownNow);
+        executor = listeningDecorator(coreExecutor);
+    }
+
+    public void initializeTempStorage()
+            throws IOException
+    {
+        tempStorage.initialize();
+    }
+
+    @Override
+    public OutputBufferInfo getInfo()
+    {
+        return new OutputBufferInfo(
+                "SPOOLING",
+                state.get(),
+                state.get().canAddBuffers(),
+                state.get().canAddPages(),
+                totalBufferedBytes.get(),
+                totalBufferedPages.get(),
+                totalRowsAdded.get(),
+                totalPagesAdded.get(),
+                ImmutableList.of()
+        );
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return state.get() == BufferState.FINISHED;
+    }
+
+    @Override
+    public double getUtilization()
+    {
+        return 0;
+    }
+
+    @Override
+    public boolean isOverutilized()
+    {
+        return false;
+    }
+
+    @Override
+    public long getPeakMemoryUsage()
+    {
+        return 0;
```

Comment:
More of a question but since this has memory component as well (besides disk) would it be better to track it here as well?

Reply:
Good point, I'll change it.